File: cpp/src/broker/NullMessageStore.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

#include "NullMessageStore.h"

#include "RecoveryManager.h"

#include <iostream>

using namespace qpid::broker;

NullMessageStore::NullMessageStore(bool _warn) : warn(_warn){}

void NullMessageStore::create(const PersistableQueue& queue)
{
    if (warn) std::cout << "WARNING: Can't create durable queue '" << queue.getName() << "'. Persistence not enabled." << std::endl;
}

void NullMessageStore::destroy(const PersistableQueue& queue)
{
    if (warn) std::cout << "WARNING: Can't destroy durable queue '" << queue.getName() << "'. Persistence not enabled." << std::endl;
}

void NullMessageStore::create(const PersistableExchange&)
{
}

void NullMessageStore::destroy(const PersistableExchange&)
{
}

void NullMessageStore::recover(RecoveryManager&)
{
    if (warn) std::cout << "WARNING: Persistence not enabled, no recovery of queues or messages." << std::endl;
}

void NullMessageStore::stage(PersistableMessage&)
{
    if (warn) std::cout << "WARNING: Can't stage message. Persistence not enabled." << std::endl;
}

void NullMessageStore::destroy(PersistableMessage&)
{
    if (warn) std::cout << "WARNING: No need to destroy staged message. Persistence not enabled." << std::endl;
}

void NullMessageStore::appendContent(PersistableMessage&, const string&)
{
    if (warn) std::cout << "WARNING: Can't append content. Persistence not enabled." << std::endl;
}

void NullMessageStore::loadContent(PersistableMessage&, string&, uint64_t, uint32_t)
{
    if (warn) std::cout << "WARNING: Can't load content. Persistence not enabled." << std::endl;
}

void NullMessageStore::enqueue(TransactionContext*, PersistableMessage&, const PersistableQueue& queue)
{
    if (warn) std::cout << "WARNING: Can't enqueue message onto '" << queue.getName() << "'. Persistence not enabled." << std::endl;
}

void NullMessageStore::dequeue(TransactionContext*, PersistableMessage&, const PersistableQueue& queue)
{
    if (warn) std::cout << "WARNING: Can't dequeue message from '" << queue.getName() << "'. Persistence not enabled." << std::endl;
}

std::auto_ptr<TransactionContext> NullMessageStore::begin()
{
    return std::auto_ptr<TransactionContext>();
}

std::auto_ptr<TPCTransactionContext> NullMessageStore::begin(const std::string&)
{
    return std::auto_ptr<TPCTransactionContext>();
}

void NullMessageStore::prepare(TPCTransactionContext&)
{
}

void NullMessageStore::commit(TransactionContext&)
{
}

void NullMessageStore::abort(TransactionContext&)
{
}

File: java/broker-core/src/main/java/org/apache/qpid/server/store/NullMessageStore.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.server.store;

import java.util.UUID;

import org.apache.qpid.server.model.ConfiguredObject;
import org.apache.qpid.server.store.handler.ConfiguredObjectRecordHandler;
import org.apache.qpid.server.store.handler.DistributedTransactionHandler;
import org.apache.qpid.server.store.handler.MessageHandler;
import org.apache.qpid.server.store.handler.MessageInstanceHandler;

public abstract class NullMessageStore implements MessageStore, DurableConfigurationStore, MessageStoreProvider
{
    private ConfiguredObjectRecord[] _initialRecords;

    @Override
    public MessageStore getMessageStore()
    {
        return this;
    }

    @Override
    public void openConfigurationStore(ConfiguredObject<?> parent,
                                       final boolean overwrite,
                                       final ConfiguredObjectRecord... initialRecords)
    {
        _initialRecords = initialRecords;
    }

    @Override
    public void update(boolean createIfNecessary, ConfiguredObjectRecord... records)
    {
    }

    @Override
    public UUID[] remove(final ConfiguredObjectRecord... objects)
    {
        final UUID[] removed = new UUID[objects.length];
        for(int i = 0; i < objects.length; i++)
        {
            removed[i] = objects[i].getId();
        }
        return removed;
    }

    @Override
    public void create(ConfiguredObjectRecord record)
    {
    }

    @Override
    public void openMessageStore(ConfiguredObject<?> parent)
    {
    }

    @Override
    public void upgradeStoreStructure() throws StoreException
    {

    }

    @Override
    public void closeMessageStore()
    {
    }

    @Override
    public void closeConfigurationStore()
    {
    }

    @Override
    public <T extends StorableMessageMetaData> StoredMessage<T> addMessage(T metaData)
    {
        return null;
    }

    @Override
    public boolean isPersistent()
    {
        return false;
    }

    @Override
    public Transaction newTransaction()
    {
        return null;
    }

    @Override
    public void addEventListener(EventListener eventListener, Event... events)
    {
    }

    @Override
    public String getStoreLocation()
    {
        return null;
    }

    @Override
    public void onDelete()
    {
    }

    @Override
    public void visitConfiguredObjectRecords(ConfiguredObjectRecordHandler handler) throws StoreException
    {
        handler.begin();
        if(_initialRecords != null)
        {
            for(ConfiguredObjectRecord record : _initialRecords)
            {
                if(!handler.handle(record))
                {
                    break;
                }
            }
        }
        handler.end();
    }

    @Override
    public void visitMessages(MessageHandler handler) throws StoreException
    {
    }

    @Override
    public void visitMessageInstances(TransactionLogResource queue, MessageInstanceHandler handler) throws StoreException
    {
    }

    @Override
    public void visitMessageInstances(MessageInstanceHandler handler) throws StoreException
    {
    }

    @Override
    public void visitDistributedTransactions(DistributedTransactionHandler handler) throws StoreException
    {
    }

    @Override
    public long getNextMessageId()
    {
        return 0;
    }

    @Override
    public StoredMessage<?> getMessage(final long messageId)
    {
        return null;
    }
}
